apps/agent/test/agent_shared.js

```diff
@@ -94,6 +100,144 @@ module.exports = (
+            const cheapFallbackTarget = await DestinationMock.new(false)
+            const receipt = await agent.safeExecute(cheapFallbackTarget.address, noData, { from: authorized })
+
+            assertAmountOfEvents(receipt, 'SafeExecute')
+          })
+
+          it('it can execute expensive fallback actions', async () => {
+            const expensiveFallbackTarget = await DestinationMock.new(true)
+            assert.equal(await expensiveFallbackTarget.counter(), 0)
+            const receipt = await agent.safeExecute(expensiveFallbackTarget.address, noData, { from: authorized })
+
+            assertAmountOfEvents(receipt, 'SafeExecute')
+            assert.equal(await expensiveFallbackTarget.counter(), 1) // fallback increments counter
+          })
+
+          it('it can execute with data when target is not a contract', async () => {
+            const nonContract = accounts[8] // random account
+            const randomData = '0x12345678'
+            const receipt = await agent.safeExecute(nonContract, randomData, { from: authorized })
+
+            assertAmountOfEvents(receipt, 'SafeExecute')
+          })
+
+          it('it can execute without data when target is not a contract', async () => {
+            const nonContract = accounts[8] // random account
+            const receipt = await agent.safeExecute(nonContract, noData, { from: authorized })
+
+            assertAmountOfEvents(receipt, 'SafeExecute')
+          })
+
+          it('it can forward success return data', async () => {
+            const { to, data } = encodeFunctionCall(target, 'execute')
+
+            // We make a call to easily get what data could be gotten inside the EVM
+            // Contract -> agent.safeExecute -> Target.func (would allow Contract to have access to this data)
+            const call = encodeFunctionCall(agent, 'safeExecute', to, data, { from: authorized })
+            const returnData = await web3Call(call)
+
+            // ExecutionTarget.execute() increments the counter by 1
+            assert.equal(ethABI.decodeParameter('uint256', returnData), 1)
+          })
+
+          it('it should revert if executed action reverts', async () => {
+            // TODO: Check revert data was correctly forwarded
+            // ganache currently doesn't support fetching this data
+            const data = target.contract.fail.getData()
+            await assertRevert(() => agent.safeExecute(target.address, data, { from: authorized }))
+          })
+        })
+
+        context('> but target is a protected ERC20', () => {
+          it('it should revert', async () => {
+            const approve = token1.contract.approve.getData(target.address, 10)
+
+            await assertRevert(() => agent.safeExecute(token1.address, approve, { from: authorized }))
+          })
+        })
+
+        context('> and target is not a protected ERC20 but action affects a protected ERC20 balance', () => {
+          it('it should revert', async () => {
```

Comment:
We should also test the execution modifying the protected token list which should cause a revert.